### PR TITLE
UCP/WIREUP/AM-BW: mask lanes by device

### PR DESCRIPTION
**Pull request by hoopoepg**
- mask AM BW lanes by device index
- added remote address mask to select transport routine

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5933/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe should I make similar changes to RMA BW?

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -129,7 +129,8 @@ static int ucp_wireup_is_reachable(ucp_worker_h worker, ucp_rsc_index_t rsc_inde
 static UCS_F_NOINLINE ucs_status_t
 ucp_wireup_select_transport(ucp_ep_h ep, const ucp_address_entry_t *address_list,
                             unsigned address_count, const ucp_wireup_criteria_t *criteria,
-                            uint64_t tl_bitmap, uint64_t remote_md_map, int show_error,
+                            uint64_t tl_bitmap, uint64_t remote_md_map,
+                            uint64_t remote_addr_bitmap, int show_error,
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: wouldn't be simpler to pass remote_dev_bitmap and local_dev_bitmap?
so will not need XX_unset_by_device

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: it will require more global changes in transport selection routine, ok, will do

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5943/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe more comments?

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -161,12 +162,15 @@ ucp_wireup_select_transport(ucp_ep_h ep, const ucp_address_entry_t *address_list
     addr_index_map = 0;
     for (ae = address_list; ae < address_list + address_count; ++ae) {
         addr_index = ae - address_list;
-        if (!(remote_md_map & UCS_BIT(ae->md_index))) {
+        if (!(remote_dev_bitmap & UCS_BIT(ae->dev_index))) {
+            ucs_trace("addr[%d]: not in use, because device[%d]",
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: because on device[%d]

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -877,20 +888,19 @@ static ucs_status_t ucp_wireup_add_am_bw_lanes(ucp_ep_h ep, const ucp_ep_params_
...
     for (lane_desc_idx = 0; lane_desc_idx < *num_lanes_p; ++lane_desc_idx) {
         if (lane_descs[lane_desc_idx].usage & UCP_WIREUP_LANE_USAGE_AM) {
-            addr_index     = lane_descs[lane_desc_idx].addr_index;
-            rsc_index      = lane_descs[lane_desc_idx].rsc_index;
-            remote_md_map &= ~UCS_BIT(address_list[addr_index].md_index);
-            md_map        |= UCS_BIT(context->tl_rscs[rsc_index].md_index);
-            tl_bitmap      = ucp_wireup_unset_tl_by_md(ep, tl_bitmap, rsc_index);
+            addr_index        = lane_descs[lane_desc_idx].addr_index;
+            rsc_index         = lane_descs[lane_desc_idx].rsc_index;
+            md_map           |= UCS_BIT(context->tl_rscs[rsc_index].md_index);
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: i guess we don't need to deselect by md_map now?
can you pls test that it works on a machine with dual-port HCA such as ConnectX-3 or Connect-IB?

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: fixed

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: md_map here is used to limit number of lanes by MD's
tested on Connect-X3 (single HCA with double port) - number of lanes has no effect to results (single lane is always selected)

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: correction:
ucx_info say:
  lane[0]: 3:rc/mlx4_0:1 md[1]           -> md[1] rma_bw#0 am am_bw#0
  lane[1]: 4:rc/mlx4_0:2 md[1]           -> md[1] rma_bw#1 am_bw#1 wireup{ud/mlx4_0:1}

it seems different ports have different device numbers

Comment (yosefe) [reply src/ucp/wireup/select.c]: connect-X3 is PCIe 8x so should not have large difference with FDR
in order to test can degrade port speed to QDR

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5947/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5949/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe last commit fixes wireup non-symmetric issue

**Pull request merged**